File: Codigos_ArrietaFCII/descomposicionLU.py
import numpy as np
import time

##Descompocicion LU para un sistema lineal de ecuaciones
# Matriz A
A = np.array([
    [2.0, -1.0, 1.0],
    [3.0, 3.0, 9.0],
    [3.0, 3.0, 5.0]
])

# Vector b
b = np.array([2.0, -1.0, 4.0])

N = len(A)
inicio= time.perf_counter()
# Matrices L y U
L = np.zeros((N,N))
U = np.zeros((N,N))


# hacemos la descompocicion LU


for i in range(N):

    # Calcular U
    for j in range(i, N):
        suma = 0
        for k in range(i):
            suma += L[i,k] * U[k,j]

        U[i,j] = A[i,j] - suma

    # Calcular L
    for j in range(i, N):

        if i == j:
            L[i,i] = 1
        else:
            suma = 0
            for k in range(i):
                suma += L[j,k] * U[k,i]

            L[j,i] = (A[j,i] - suma) / U[i,i]



# SUSTITUCION HACIA ADELANTE
# Ly = b


y = np.zeros(N)

for i in range(N):

    suma = 0
    for j in range(i):
        suma += L[i,j] * y[j]

    y[i] = b[i] - suma



# sustitucion inversa
# Ux = y


x = np.zeros(N)

for i in range(N-1, -1, -1):

    suma = 0
    for j in range(i+1, N):
        suma += U[i,j] * x[j]

    x[i] = (y[i] - suma) / U[i,i]


print("Matriz L:")
print(L)

print("\nMatriz U:")
print(U)

print("\nVector y:")
print(y)

print("\nSolucion x:")
print(x)

fin = time.perf_counter()
print("\ntime")
print((fin-inicio)*1e6)

File: Codigos_ArrietaFCII/Rust_Arrieta/src/DescomposicionLU.rs
use std::time::Instant;

fn main() {

    // Matriz A
    let a: Vec<Vec<f64>> = vec![
        vec![2.0, -1.0, 1.0],
        vec![3.0, 3.0, 9.0],
        vec![3.0, 3.0, 5.0],
    ];

    // Vector b
    let b: Vec<f64> = vec![2.0, -1.0, 4.0];

    let n = a.len();

    let inicio = Instant::now();

    // Matrices L y U
    let mut l = vec![vec![0.0; n]; n];
    let mut u = vec![vec![0.0; n]; n];

    // Descomposición LU
    for i in 0..n {

        // Calcular U
        for j in i..n {

            let mut suma = 0.0;

            for k in 0..i {
                suma += l[i][k] * u[k][j];
            }

            u[i][j] = a[i][j] - suma;
        }

        // Calcular L
        for j in i..n {

            if i == j {
                l[i][i] = 1.0;
            } else {

                let mut suma = 0.0;

                for k in 0..i {
                    suma += l[j][k] * u[k][i];
                }

                l[j][i] = (a[j][i] - suma) / u[i][i];
            }
        }
    }

    // Sustitución hacia adelante
    // Ly = b

    let mut y = vec![0.0; n];

    for i in 0..n {

        let mut suma = 0.0;

        for j in 0..i {
            suma += l[i][j] * y[j];
        }

        y[i] = b[i] - suma;
    }

    // Sustitución hacia atrás
    // Ux = y

    let mut x = vec![0.0; n];

    for i in (0..n).rev() {

        let mut suma = 0.0;

        for j in (i + 1)..n {
            suma += u[i][j] * x[j];
        }

        x[i] = (y[i] - suma) / u[i][i];
    }

    // Resultados

    println!("Matriz L:");
    for fila in &l {
        println!("{:?}", fila);
    }

    println!("\nMatriz U:");
    for fila in &u {
        println!("{:?}", fila);
    }

    println!("\nVector y:");
    println!("{:?}", y);

    println!("\nSolucion sistema x:");
    println!("{:?}", x);

    let duracion = inicio.elapsed();

    println!("\nTiempo (microsegundos):");
    println!("{}", duracion.as_micros());
}